File: services/entity-resolution-service/src/fusion_base/domain/engine/blocking.rs
use std::collections::{BTreeMap, BTreeSet};

use crate::models::{cluster::EntityRecord, match_rule::BlockingStrategyConfig};

use super::comparator::normalize_text;

#[derive(Debug, Clone)]
pub struct CandidatePair {
    pub left: EntityRecord,
    pub right: EntityRecord,
    pub blocking_key: String,
}

pub fn build_candidate_pairs(
    records: &[EntityRecord],
    strategy: &BlockingStrategyConfig,
) -> Vec<CandidatePair> {
    match strategy.strategy_type.as_str() {
        "sorted-neighborhood" => sorted_neighborhood_pairs(records, strategy),
        "lsh" => lsh_pairs(records, strategy),
        _ => key_based_pairs(records, strategy),
    }
}

fn key_based_pairs(
    records: &[EntityRecord],
    strategy: &BlockingStrategyConfig,
) -> Vec<CandidatePair> {
    let mut groups = BTreeMap::<String, Vec<&EntityRecord>>::new();
    for record in records {
        groups
            .entry(blocking_key(record, &strategy.key_fields))
            .or_default()
            .push(record);
    }

    let mut emitted = BTreeSet::new();
    let mut pairs = Vec::new();

    for (key, group) in groups {
        for left_index in 0..group.len() {
            for right_index in left_index + 1..group.len() {
                emit_pair(
                    &mut pairs,
                    &mut emitted,
                    &key,
                    group[left_index],
                    group[right_index],
                );
            }
        }
    }

    pairs
}

fn sorted_neighborhood_pairs(
    records: &[EntityRecord],
    strategy: &BlockingStrategyConfig,
) -> Vec<CandidatePair> {
    let mut sorted = records
        .iter()
        .map(|record| (blocking_key(record, &strategy.key_fields), record))
        .collect::<Vec<_>>();
    sorted.sort_by(|left, right| left.0.cmp(&right.0));

    let window_size = strategy.window_size.max(2) as usize;
    let mut emitted = BTreeSet::new();
    let mut pairs = Vec::new();

    for left_index in 0..sorted.len() {
        let upper = (left_index + window_size).min(sorted.len());
        for right_index in left_index + 1..upper {
            let key = format!("{}|{}", sorted[left_index].0, sorted[right_index].0);
            emit_pair(
                &mut pairs,
                &mut emitted,
                &key,
                sorted[left_index].1,
                sorted[right_index].1,
            );
        }
    }

    pairs
}

fn lsh_pairs(records: &[EntityRecord], strategy: &BlockingStrategyConfig) -> Vec<CandidatePair> {
    let bucket_count = strategy.bucket_count.max(4) as u64;
    let mut buckets = BTreeMap::<String, Vec<&EntityRecord>>::new();

    for record in records {
        for bucket in record_buckets(record, bucket_count) {
            buckets.entry(bucket).or_default().push(record);
        }
    }

    let mut emitted = BTreeSet::new();
    let mut pairs = Vec::new();

    for (bucket, group) in buckets {
        for left_index in 0..group.len() {
            for right_index in left_index + 1..group.len() {
                emit_pair(
                    &mut pairs,
                    &mut emitted,
                    &bucket,
                    group[left_index],
                    group[right_index],
                );
            }
        }
    }

    pairs
}

fn record_buckets(record: &EntityRecord, bucket_count: u64) -> Vec<String> {
    let mut buckets = BTreeSet::new();
    let seed = normalize_text(&record.display_name);

    for token in seed.split_whitespace() {
        if token.is_empty() {
            continue;
        }

        let hash = token.bytes().fold(0_u64, |accumulator, byte| {
            accumulator.wrapping_mul(31).wrapping_add(byte as u64)
        });
        buckets.insert(format!("bucket-{}", hash % bucket_count));
    }

    if buckets.is_empty() {
        buckets.insert("bucket-0".to_string());
    }

    buckets.into_iter().collect()
}

fn emit_pair(
    pairs: &mut Vec<CandidatePair>,
    emitted: &mut BTreeSet<String>,
    blocking_key: &str,
    left: &EntityRecord,
    right: &EntityRecord,
) {
    let pair_key = pair_key(&left.record_id, &right.record_id);
    if !emitted.insert(pair_key) {
        return;
    }

    pairs.push(CandidatePair {
        left: left.clone(),
        right: right.clone(),
        blocking_key: blocking_key.to_string(),
    });
}

fn pair_key(left_id: &str, right_id: &str) -> String {
    if left_id <= right_id {
        format!("{left_id}::{right_id}")
    } else {
        format!("{right_id}::{left_id}")
    }
}

fn blocking_key(record: &EntityRecord, key_fields: &[String]) -> String {
    let mut parts = key_fields
        .iter()
        .filter_map(|field| extract_field(record, field))
        .map(|value| normalize_text(&value))
        .filter(|value| !value.is_empty())
        .map(|value| value.chars().take(6).collect::<String>())
        .collect::<Vec<_>>();

    if parts.is_empty() {
        parts.push(
            normalize_text(&record.display_name)
                .chars()
                .take(6)
                .collect(),
        );
    }

    parts.join("|")
}

fn extract_field(record: &EntityRecord, field: &str) -> Option<String> {
    match field {
        "display_name" | "name" => Some(record.display_name.clone()),
        _ => record
            .attributes
            .get(field)
            .and_then(|value| value.as_str())
            .map(ToOwned::to_owned),
    }
}

File: openfoundry-go/services/entity-resolution-service/internal/domain/engine/blocking.go
package engine

import (
	"fmt"
	"sort"

	"github.com/openfoundry/openfoundry-go/services/entity-resolution-service/internal/models"
)

// CandidatePair mirrors fusion_base::domain::engine::blocking::CandidatePair.
type CandidatePair struct {
	Left        models.EntityRecord
	Right       models.EntityRecord
	BlockingKey string
}

// BuildCandidatePairs ports `blocking::build_candidate_pairs`.
func BuildCandidatePairs(records []models.EntityRecord, strategy models.BlockingStrategyConfig) []CandidatePair {
	switch strategy.StrategyType {
	case "sorted-neighborhood":
		return sortedNeighborhoodPairs(records, strategy)
	case "lsh":
		return lshPairs(records, strategy)
	default:
		return keyBasedPairs(records, strategy)
	}
}

func keyBasedPairs(records []models.EntityRecord, strategy models.BlockingStrategyConfig) []CandidatePair {
	groups := map[string][]int{}
	for i, record := range records {
		key := blockingKey(record, strategy.KeyFields)
		groups[key] = append(groups[key], i)
	}

	keys := make([]string, 0, len(groups))
	for k := range groups {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	emitted := map[string]struct{}{}
	pairs := []CandidatePair{}
	for _, key := range keys {
		group := groups[key]
		for li := 0; li < len(group); li++ {
			for ri := li + 1; ri < len(group); ri++ {
				pairs = emitPair(pairs, emitted, key, records[group[li]], records[group[ri]])
			}
		}
	}
	return pairs
}

func sortedNeighborhoodPairs(records []models.EntityRecord, strategy models.BlockingStrategyConfig) []CandidatePair {
	type keyed struct {
		key string
		rec models.EntityRecord
	}
	sorted := make([]keyed, len(records))
	for i, r := range records {
		sorted[i] = keyed{key: blockingKey(r, strategy.KeyFields), rec: r}
	}
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].key < sorted[j].key })

	windowSize := int(strategy.WindowSize)
	if windowSize < 2 {
		windowSize = 2
	}
	emitted := map[string]struct{}{}
	pairs := []CandidatePair{}
	for li := 0; li < len(sorted); li++ {
		upper := li + windowSize
		if upper > len(sorted) {
			upper = len(sorted)
		}
		for ri := li + 1; ri < upper; ri++ {
			combinedKey := fmt.Sprintf("%s|%s", sorted[li].key, sorted[ri].key)
			pairs = emitPair(pairs, emitted, combinedKey, sorted[li].rec, sorted[ri].rec)
		}
	}
	return pairs
}

func lshPairs(records []models.EntityRecord, strategy models.BlockingStrategyConfig) []CandidatePair {
	bucketCount := uint64(strategy.BucketCount)
	if bucketCount < 4 {
		bucketCount = 4
	}
	buckets := map[string][]int{}
	for i, record := range records {
		for _, bucket := range recordBuckets(record, bucketCount) {
			buckets[bucket] = append(buckets[bucket], i)
		}
	}
	keys := make([]string, 0, len(buckets))
	for k := range buckets {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	emitted := map[string]struct{}{}
	pairs := []CandidatePair{}
	for _, bucket := range keys {
		group := buckets[bucket]
		for li := 0; li < len(group); li++ {
			for ri := li + 1; ri < len(group); ri++ {
				pairs = emitPair(pairs, emitted, bucket, records[group[li]], records[group[ri]])
			}
		}
	}
	return pairs
}

func recordBuckets(record models.EntityRecord, bucketCount uint64) []string {
	seed := NormalizeText(record.DisplayName)
	tokens := tokenize(seed)
	bucketSet := map[string]struct{}{}
	for _, token := range tokens {
		if token == "" {
			continue
		}
		var hash uint64
		for _, b := range []byte(token) {
			hash = hash*31 + uint64(b)
		}
		bucketSet[fmt.Sprintf("bucket-%d", hash%bucketCount)] = struct{}{}
	}
	if len(bucketSet) == 0 {
		bucketSet["bucket-0"] = struct{}{}
	}
	out := make([]string, 0, len(bucketSet))
	for k := range bucketSet {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

func tokenize(input string) []string {
	out := []string{}
	current := []rune{}
	for _, r := range input {
		switch r {
		case ' ', '\t', '\n', '\r':
			if len(current) > 0 {
				out = append(out, string(current))
				current = current[:0]
			}
		default:
			current = append(current, r)
		}
	}
	if len(current) > 0 {
		out = append(out, string(current))
	}
	return out
}

func emitPair(pairs []CandidatePair, emitted map[string]struct{}, blockingKey string, left, right models.EntityRecord) []CandidatePair {
	key := pairKey(left.RecordID, right.RecordID)
	if _, ok := emitted[key]; ok {
		return pairs
	}
	emitted[key] = struct{}{}
	return append(pairs, CandidatePair{Left: left, Right: right, BlockingKey: blockingKey})
}

func pairKey(left, right string) string {
	if left <= right {
		return left + "::" + right
	}
	return right + "::" + left
}

func blockingKey(record models.EntityRecord, keyFields []string) string {
	parts := []string{}
	for _, field := range keyFields {
		raw, ok := extractFieldString(record, field)
		if !ok {
			continue
		}
		normalized := NormalizeText(raw)
		if normalized == "" {
			continue
		}
		parts = append(parts, takeRunes(normalized, 6))
	}
	if len(parts) == 0 {
		parts = append(parts, takeRunes(NormalizeText(record.DisplayName), 6))
	}
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += "|"
		}
		out += p
	}
	return out
}

func extractFieldString(record models.EntityRecord, field string) (string, bool) {
	switch field {
	case "display_name", "name":
		return record.DisplayName, true
	default:
		v, ok := record.Attributes[field]
		if !ok {
			return "", false
		}
		s, ok := v.(string)
		if !ok {
			return "", false
		}
		return s, true
	}
}

func takeRunes(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n])
}
